### PR TITLE
Add support for different src and dst datatypes in the SYCL implementation for pooling

### PR DESCRIPTION
**Pull request by kala855**
# Description

This pull request facilitates the utilization of distinct source (src) and destination (dst) datatypes during the forward inference pass within the SYCL pooling operator implementation.

### CODE REVIEW

File: tests/benchdnn/pool/ref_pool.cpp [hunk 1]
```diff
@@ -38,6 +38,9 @@ void compute_ref_fwd(const prb_t *prb, const args_t &args) {
         // XXX: this is a hack to let tests with padded area to pass for bf16
         // dt due to the library initialize values with -max_dt, but not -INF.
         float max_value = lowest_dt(prb->dst_dt());
+        if (prb->src_dt() == dnnl_u8 || prb->src_dt() == dnnl_s8
+                || prb->src_dt() == dnnl_f16 || prb->src_dt() == dnnl_f32)
+            max_value = lowest_dt(prb->src_dt());
```
Comment (dzarukin) [inline tests/benchdnn/pool/ref_pool.cpp:43]: Could you elaborate, please, why is this needed?

File: src/gpu/sycl/ref_pooling.hpp [hunk 2]
```diff
@@ -53,18 +53,13 @@ struct ref_pooling_fwd_t : public sycl_gpu_primitive_t {
...
-                            || utils::everyone_is(bf16, src_md(0)->data_type,
-                                    dst_md(0)->data_type)
-                            || utils::everyone_is(f16, src_md(0)->data_type,
-                                    dst_md(0)->data_type)
-                            || utils::everyone_is(s32, src_md(0)->data_type,
-                                    dst_md(0)->data_type))
+                    && (!utils::one_of(f64, src_md(0)->data_type)
+                            && !utils::one_of(f64, dst_md(0)->data_type))
+                    && (IMPLICATION(utils::one_of(bf16, src_md(0)->data_type),
+                            utils::one_of(bf16, dst_md(0)->data_type)))
```
Comment (dzarukin) [inline src/gpu/sycl/ref_pooling.hpp:59]: Could you explain, please, what makes bf16 so special?

Comment (kala855) [reply src/gpu/sycl/ref_pooling.hpp:59]: In this instance, we're adhering to the specifications outlined in the table available [here](https://oneapi-src.github.io/oneDNN/dev_guide_pooling.html#data-type-support). Notably, the `bf16` scenario stands out as one case requiring identical source and destination data types. Thanks for the comment. Let me know if something else needs to be clarified.

Comment (dzarukin) [reply src/gpu/sycl/ref_pooling.hpp:59]: I don't see a strong reason to limit the support if it can work as expected.

File: tests/benchdnn/pool/ref_pool.cpp (hunk 1 above)
Comment (dzarukin) [reply tests/benchdnn/pool/ref_pool.cpp:43]: Different data types are supported for other backends and yet so far everything worked beside that issue.
Please post benchdnn command and output without this patch to benchdnn where it fails. 2-3 would be enough. Thanks.

Comment (dzarukin) [reply tests/benchdnn/pool/ref_pool.cpp:43]: @kala855, thanks for these examples. It seems there's no clear option how to proceed with it, and it will require clarification with users on what's the expected behavior here. Two possible situations:
1. In the sense that there's no quantized operations defined in framework graphs, all operations are considered as f32/bf16. It means that data would be converted to f32/bf16, computed, and then converted to output data type separately. This is aligned with current benchdnn expectations.
2. The quantized operation uses src_dt as a data type to update values for max pooling. Then converts it into f32 to apply post-ops, if any, and then converts to dst_dt. This is aligned with the new change and breaking the expectations of the previous one.

A request here is to postpone this PR until we get clarifications from frameworks and proceed with the most reasonable option. Thanks.

**Comment by kala855** [comment]
@dzarukin We take into account what was discussed during the meeting. The PR now avoids applying pooling src:dst (u8:u8, s8:s8).

File: src/gpu/sycl/ref_pooling.hpp [hunk 3]
```diff
@@ -53,18 +53,19 @@ struct ref_pooling_fwd_t : public sycl_gpu_primitive_t {
...
-                            || utils::everyone_is(f16, src_md(0)->data_type,
-                                    dst_md(0)->data_type)
-                            || utils::everyone_is(s32, src_md(0)->data_type,
-                                    dst_md(0)->data_type))
+                    && (IMPLICATION(utils::one_of(bf16, src_md(0)->data_type),
+                            utils::one_of(bf16, dst_md(0)->data_type)))
+                    && f64 != src_md(0)->data_type
+                    && f64 != dst_md(0)->data_type
+                    && (IMPLICATION(bf16 == src_md(0)->data_type,
+                            bf16 == dst_md(0)->data_type))
```
Comment (mgouicem) [inline src/gpu/sycl/ref_pooling.hpp:61]: this is redundant with line 56-57.

File: src/gpu/sycl/ref_pooling.hpp [hunk 4]
```diff
@@ -53,18 +53,19 @@ struct ref_pooling_fwd_t : public sycl_gpu_primitive_t {
...
+                    && (IMPLICATION(src_md(0)->data_type == s8,
+                            dst_md(0)->data_type != s8))
+                    && (IMPLICATION(src_md(0)->data_type == u8,
+                            dst_md(0)->data_type != u8))
+                    && (IMPLICATION(
+                            src_md(0)->data_type != dst_md(0)->data_type,
+                            desc_.prop_kind == forward_inference))
```
Comment (mgouicem) [inline src/gpu/sycl/ref_pooling.hpp:68]: Not sure I understand the logic here. You are disabling src and dst datatype to be the same except for bf16.
Having the same datatype for src and dst should be valid.

**Comment by mgouicem** [comment]
> @dzarukin We take into account what was discussed during the meeting. The PR now avoids applying pooling src:dst (u8:u8, s8:s8).

Could you clarify what issue does this fix?
IIUC, the main problem is in the selection of datatype for the max value computation (see [this comment](https://github.com/oneapi-src/oneDNN/pull/1878#discussion_r1583606634)). In the case where both src and dst have same datatype, there should be no ambiguity.

File: src/gpu/sycl/ref_pooling.hpp (hunk 3 above)
Comment (kala855) [reply src/gpu/sycl/ref_pooling.hpp]: It was fixed in the current commit. Thanks.

File: src/gpu/sycl/ref_pooling.hpp (hunk 4 above)
Comment (kala855) [reply src/gpu/sycl/ref_pooling.hpp:66]: In this [table](https://oneapi-src.github.io/oneDNN/dev_guide_pooling.html#data-type-support) src/dst datatypes could be different just when `forward_inference` is used. The bf16 dt is not combined with any other dt. Maybe I am misunderstanding something in the table. Let me know if this clarifies a little bit the questions. Thanks.

**Comment by kala855** [comment]
> > @dzarukin We take into account what was discussed during the meeting. The PR now avoids applying pooling src:dst (u8:u8, s8:s8).
> 
> Could you clarify what issue does this fix? IIUC, the main problem is in the selection of datatype for the max value computation (see [this comment](https://github.com/oneapi-src/oneDNN/pull/1878#discussion_r1583606634)). In the case where both src and dst have same datatype, there should be no ambiguity.

@mgouicem We received feedback mentioning that we need to avoid the use of `u8:u8` and `s8:s8`. Maybe @dzarukin have any additional clue about it. If this combination is accepted we need to enabled it an run some tests. I am in line with your thoughts about the max value computation.

**Comment by dzarukin** [comment]
> @mgouicem We received feedback mentioning that we need to avoid the use of `u8:u8` and `s8:s8`. Maybe @dzarukin have any additional clue about it. If this combination is accepted we need to enabled it an run some tests. I am in line with your thoughts about the max value computation.

u8:s8 and s8:u8 have semantic problems.

**Comment by kala855** [comment]
> > @mgouicem We received feedback mentioning that we need to avoid the use of `u8:u8` and `s8:s8`. Maybe @dzarukin have any additional clue about it. If this combination is accepted we need to enabled it an run some tests. I am in line with your thoughts about the max value computation.
> 
> u8:s8 and s8:u8 have semantic problems.

@dzarukin just to be sure ... The ones we will need to avoid are **u8:s8** and **s8:u8** ... However, **u8:u8** and **s8:s8** are allowed?
The previous just to be sure we are talking about the same and proceed to do the corrections. Thank you.

**Comment by dzarukin** [comment]
> > > @mgouicem We received feedback mentioning that we need to avoid the use of `u8:u8` and `s8:s8`. Maybe @dzarukin have any additional clue about it. If this combination is accepted we need to enabled it an run some tests. I am in line with your thoughts about the max value computation.
> > 
> > 
> > u8:s8 and s8:u8 have semantic problems.
> 
> @dzarukin just to be sure ... The ones we will need to avoid are **u8:s8** and **s8:u8** ... However, **u8:u8** and **s8:s8** are allowed? The previous just to be sure we are talking about the same and proceed to do the corrections. Thank you.

Yes, those are allowed and should be supported. There's no ambiguity in such combinations - same init value is used on both ends.

**Comment by kala855** [comment]
> > > > @mgouicem We received feedback mentioning that we need to avoid the use of `u8:u8` and `s8:s8`. Maybe @dzarukin have any additional clue about it. If this combination is accepted we need to enabled it an run some tests. I am in line with your thoughts about the max value computation.
> > > 
> > > 
> > > u8:s8 and s8:u8 have semantic problems.
> > 
> > 
> > @dzarukin just to be sure ... The ones we will need to avoid are **u8:s8** and **s8:u8** ... However, **u8:u8** and **s8:s8** are allowed? The previous just to be sure we are talking about the same and proceed to do the corrections. Thank you.
> 
> Yes, those are allowed and should be supported. There's no ambiguity in such combinations - same init value is used on both ends.

The implementation now does what is suggested. Let me know if something else is missing @dzarukin . Thank you.

File: tests/benchdnn/pool/ref_pool.cpp (hunk 1 above)
Comment (dzarukin) [reply tests/benchdnn/pool/ref_pool.cpp:43]: I'd expect benchdnn change to be dropped from the final version with limited int8 support...

File: src/gpu/generic/sycl/ref_pooling.hpp [hunk 5]
```diff
@@ -53,18 +53,17 @@ struct ref_pooling_fwd_t : public gpu::generic::sycl::primitive_t {
...
-                            || utils::everyone_is(f32, src_md(0)->data_type,
-                                    dst_md(0)->data_type)
-                            || utils::everyone_is(bf16, src_md(0)->data_type,
-                                    dst_md(0)->data_type)
-                            || utils::everyone_is(f16, src_md(0)->data_type,
-                                    dst_md(0)->data_type)
-                            || utils::everyone_is(s32, src_md(0)->data_type,
-                                    dst_md(0)->data_type))
+                    && (!utils::one_of(f64, src_md(0)->data_type)
+                            && !utils::one_of(f64, dst_md(0)->data_type))
```
Comment (dzarukin) [inline src/gpu/generic/sycl/ref_pooling.hpp:57]:
```suggestion
                    && (!utils::one_of(f64, src_md(0)->data_type, dst_md(0)->data_type))
```

File: src/gpu/generic/sycl/ref_pooling.hpp [hunk 6]
```diff
@@ -53,18 +53,17 @@ struct ref_pooling_fwd_t : public gpu::generic::sycl::primitive_t {
...
+                    && (IMPLICATION(utils::one_of(bf16, src_md(0)->data_type),
+                            utils::one_of(bf16, dst_md(0)->data_type)))
+                    && (IMPLICATION(src_md(0)->data_type == s8,
+                            dst_md(0)->data_type != u8))
+                    && (IMPLICATION(src_md(0)->data_type == u8,
+                            dst_md(0)->data_type != s8))
+                    && (IMPLICATION(
+                            src_md(0)->data_type != dst_md(0)->data_type,
+                            desc_.prop_kind == forward_inference))
```
Comment (dzarukin) [inline src/gpu/generic/sycl/ref_pooling.hpp:66]:
```suggestion
                            desc()->prop_kind == forward_inference))
```

File: src/gpu/generic/sycl/ref_pooling.hpp [hunk 7]
```diff
@@ -53,18 +53,17 @@ struct ref_pooling_fwd_t : public gpu::generic::sycl::primitive_t {
...
-                            || utils::everyone_is(bf16, src_md(0)->data_type,
-                                    dst_md(0)->data_type)
-                            || utils::everyone_is(f16, src_md(0)->data_type,
-                                    dst_md(0)->data_type)
-                            || utils::everyone_is(s32, src_md(0)->data_type,
-                                    dst_md(0)->data_type))
+                    && (!utils::one_of(f64, src_md(0)->data_type)
+                            && !utils::one_of(f64, dst_md(0)->data_type))
+                    && (IMPLICATION(utils::one_of(bf16, src_md(0)->data_type),
+                            utils::one_of(bf16, dst_md(0)->data_type)))
```
Comment (dzarukin) [inline src/gpu/generic/sycl/ref_pooling.hpp:59]: I'd stick to plain comparison.

File: src/gpu/generic/sycl/ref_pooling.hpp (hunk 5 above)
Comment (kala855) [reply src/gpu/generic/sycl/ref_pooling.hpp]: Change added.

File: src/gpu/generic/sycl/ref_pooling.hpp (hunk 6 above)
Comment (kala855) [reply src/gpu/generic/sycl/ref_pooling.hpp]: Change added.

File: src/gpu/generic/sycl/ref_pooling.hpp (hunk 7 above)
Comment (kala855) [reply src/gpu/generic/sycl/ref_pooling.hpp]: Done

**Review by t4c1 — APPROVED** [review]

File: tests/benchdnn/pool/ref_pool.cpp (hunk 1 above)
Comment (kala855) [reply tests/benchdnn/pool/ref_pool.cpp]: I did take into account the suggestions made. However, with limited int8 support, this part of the code still needs to be updated. Because we accept different src/dst datatypes we need to initialize the max pooling algorithm differently for Nvidia GPU. I included a conditional here to initialize the data using `prb->src_dt()`.

**Review by ShanoToni — APPROVED** [review]

**Review by densamoilov — APPROVED** [review]
